clarity/src/vm/database/clarity_db.rs

```diff
@@ -53,15 +53,209 @@ use stacks_common::consts::{
     BITCOIN_REGTEST_FIRST_BLOCK_HASH, BITCOIN_REGTEST_FIRST_BLOCK_HEIGHT,
     BITCOIN_REGTEST_FIRST_BLOCK_TIMESTAMP, FIRST_BURNCHAIN_CONSENSUS_HASH, FIRST_STACKS_BLOCK_HASH,
 };
+
+use stacks_common::address::AddressHashMode;
 use stacks_common::types::chainstate::ConsensusHash;
+use stacks_common::types::Address;
 
 use super::clarity_store::SpecialCaseHandler;
 use super::key_value_wrapper::ValueResult;
 
+use serde_json;
+
 pub const STORE_CONTRACT_SRC_INTERFACE: bool = true;
 
 pub type StacksEpoch = GenericStacksEpoch<ExecutionCost>;
 
+/// A PoX address as seen by the .pox and .pox-2 contracts.
+/// Used by the sortition DB and chains coordinator to extract addresses from the PoX contract to
+/// build the reward set and to validate block-commits.
+#[derive(Debug, PartialEq, PartialOrd, Ord, Clone, Hash, Eq, Serialize, Deserialize)]
+pub enum PoxAddress {
+    /// represents { version: (buff u1), hashbytes: (buff 20) }.
+    /// The address hash mode is optional because if we decode a legacy bitcoin address, we won't
+    /// be able to determine the hash mode since we can't distinguish segwit-p2sh from p2sh
+    Standard(StacksAddress, Option<AddressHashMode>),
+}
```

Comment:
I think this struct should be in the `chainstate` module, rather than in the Clarity module. I know there's some leakage of PoX into the Clarity module (mostly in the function signatures in the database), but most of these methods seem more important to the chainstate than it is to Clarity -- all Clarity really needs is the tuples:

```rust
/// Returns the paid out PoX addresses (as Clarity tuples of type `{ hashbytes: (buff 1), version: (buff 20) }` )
pub fn get_pox_payout_addrs_for_burnchain_height(
        &mut self,
        burnchain_block_height: u32,
    ) -> Option<(Vec<TupleData>, u128)> {
```

Whereas the chainstate module actually uses most of the methods of this struct.

Reply:
Sure, I can do that

Reply:
Done: https://github.com/stacks-network/stacks-blockchain/pull/3245/commits/49dea6ed734af9c43d257e0424e1dee135ccdf86